addon-test-support/-private/ember-exam-qunit-test-loader.js

```diff
@@ -0,0 +1,154 @@
+    super();
+    this._testModules = [];
+    this._testem = testem;
+    this._qunit = qunit;
+    this._urlParams = urlParams || getUrlParams();
+  }
+
+  get urlParams() {
+    return this._urlParams;
+  }
+
+  /**
+   * ember-cli-test-loader instantiates a new TestLoader instance and calls loadModules.
+   * EmberExamQUnitTestLoader does not support load() in favor of loadModules().
+   */
+  static load() {
+    throw new Error('`EmberExamQUnitTestLoader` doesn\'t support `load()`.');
+  }
+
+  /**
+   * require() collects the full list of modules before requiring each module with
+   * super.require(), instead of requiring and unseeing a module when each gets loaded.
+   *
+   * @param {string} moduleName
+   */
+  require(moduleName) {
+    this._testModules.push(moduleName);
+  }
+
+  /**
+   * Make unsee a no-op to avoid any unwanted resets
+   */
+  unsee() {}
+
+  /**
+   * Loads the test modules depending on the urlParam
+   */
+  loadModules() {
+    const loadBalance = this._urlParams.get('loadBalance');
+    const browserId = this._urlParams.get('browser');
+    let partitions = this._urlParams.get('partition');
+    let split = parseInt(this._urlParams.get('split'), 10);
+
+    split = isNaN(split) ? 1 : split;
+
+    if (partitions === undefined) {
+      partitions = [1];
+    } else if (!Array.isArray(partitions)) {
+      partitions = [partitions];
+    }
+
+    super.loadModules();
+
+    if (loadBalance && this._testem) {
+      this.setupLoadBalanceHandlers();
+      this._testModules = splitTestModules(
+        weightTestModules(this._testModules),
+        split,
+        partitions
+      );
+      this._testem.emit(
+        'testem:set-modules-queue',
+        this._testModules,
+        browserId
+      );
+    } else {
+      this._testModules = splitTestModules(
+        this._testModules,
+        split,
+        partitions
+      );
+      this._testModules.forEach((moduleName) => {
+        super.require(moduleName);
+        super.unsee(moduleName);
+      });
+    }
+  }
+
+  /**
+   * Allow loading one module at a time.
+   *
+   * @param {string} moduleName
+   */
+  loadIndividualModule(moduleName) {
+    if (moduleName === undefined) {
+      throw new Error(
+        'Failed to load a test module. `moduleName` is undefined in `loadIndividualModule`.'
+      );
+    }
+    super.require(moduleName);
+    super.unsee(moduleName);
+  }
+
+  /**
+   * setupLoadBalanceHandlers() registers QUnit callbacks needed for the load-balance option.
+   */
+  setupLoadBalanceHandlers() {
+    // nextModuleAsyncIterator handles the async testem events
+    // it returns an element of {value: <moduleName>, done: boolean}
+    const nextModuleAsyncIterator = new AsyncIterator(this._testem, {
+      request: 'testem:next-module-request',
+      response: 'testem:next-module-response'
+    });
+
+    const nextModuleHandler = () => {
+      return new Promise((resolve, reject) => {
```

Comment:
nit-pick: this outer promise is redundant, and requires additional potentially error prone `resolve` / `reject` wiring.

Also, the reject(`message ${err}`) will loose `err`'s meta data. As such, it is better practice to only augment the message. That way, we don't unintentionally  lose debug information.

```js
const nextModuleHandler = () => {
  return asyncModuleIterator.next().then(response => {
    if (!response.done) {
      const moduleName = response.value;
      this.loadIndividualModule(moduleName);
      // if no tests were added, request the next module
      if (this._qunit.config.queue.length === 0) {
        return  nextModuleHandler();
      }
    }
    }).catch(e => {
      if (typeof e === 'object' && e !== null && typeof e.message === 'string') {
        e.message = `EmberExam: Failed to get next test module: ${e.message}`;
      } else {
        throw new Error(`EmberExam: Failed to get next test module: ${e}`);
      }
    });
};
``